src/compiler/propagation/type_propagator.cc

```diff
@@ -382,7 +400,61 @@ void TypePropagator::call_virtual(MethodTemplate* caller, TypeScope* scope, uint
   stack->drop_arguments(arity);
 }
 
-void TypePropagator::propagate_through_lambda(Method method) {
+void TypePropagator::call_block(TypeScope* scope, uint8* site, int arity) {
+  TypeStack* stack = scope->top();
+  TypeSet receiver = stack->local(arity - 1);
+  BlockTemplate* block = receiver.block();
+
+  // If we're passing too few arguments to the block, this will
+  // throw so we do not need to update the block's argument types.
+  if (arity >= block->arity()) {
+    for (int i = 1; i < block->arity(); i++) {
+      TypeSet argument = stack->local(arity - (i + 1));
+      // Merge the argument type. If the type changed, we enqueue
+      // the block's surrounding method because it has used the
+      // argument type variable. We always start at the method,
+      // so we have the full scope available to all blocks when
+      // propagating types through them.
+      block->argument(i)->merge(this, argument);
+    }
+  }
+
+  // Drop the arguments from the stack.
+  for (int i = 0; i < arity; i++) stack->pop();
+
+  // If the return type of this block changes, we enqueue the
+  // surrounding method again.
```

Comment:
Do we analyze the whole method again?
Can we not continue at the bytecode after the block call?